Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33507 Gradcheck for complex
* #33361 torch.tensor can infer complex dtype now
* **#33268 Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float**

Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

anjali411 committed:
Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float

[ghstack-poisoned]

anjali411 commented:
discussion: https://github.com/pytorch/pytorch/pull/33175

ezyang commented on torch/csrc/utils/python_scalars.h:
```diff
@@ -40,7 +40,10 @@ inline PyObject* load_scalar(void* data, at::ScalarType scalarType) {
+    case at::kComplexFloat: {
+      std::complex<float>* data_ = (std::complex<float>*)data;
+      return PyComplex_FromDoubles(data_->real(), data_->imag());
+    }
```
nit: this bugfix could have gotten a PR all on its own

apaszke reviewed (commented):
Can't we just have complex be another dtype? I'd rather avoid adding more dtype-specific methods to tensors.

ezyang commented:
I don't really have a preference one way or another, but I saw we do have `Tensor.is_floating_point` so `Tensor.is_complex` seemed like a reasonable thing to add. There are two different complex types so you are going to want some way to conveniently check if something is complex or not.

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"




[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

apaszke commented:
I agree we should have a way to check for complex types, but wouldn't it be better to make it a dtype property?

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 commented:
> I agree we should have a way to check for complex types, but wouldn't it be better to make it a dtype property?

yeah I added both

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]

anjali411 committed:
Update on "Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float"


Differential Revision: [D19907698](https://our.internmc.facebook.com/intern/diff/D19907698)

[ghstack-poisoned]
